File: typescript/src/balances.ts
import type { Transport } from "./transport.js";

/**
 * Per-chain USDC balance snapshot returned by `/api/balances/by-chain`.
 * Matches the wire shape exactly — string balances are kept as strings
 * to avoid float-precision drift on the SDK side.
 */
export interface ChainBalance {
  chainId: number;
  chainName: string;
  usdcBalance: string;
}

interface BalancesByChainResponse {
  balances: ChainBalance[];
}

/**
 * Fetch the org treasury's per-chain USDC balance. Used by `canopy.fetch()`
 * when a 402 advertises ≥2 candidate rails to pick the funded one.
 *
 * Returns null on failure (e.g. treasury not provisioned, transient RPC
 * outage on every chain). Callers fall back to "first match" when the
 * helper returns null — better to attempt the payment and let `/api/sign`
 * surface a clear error than to wedge the request on a balance lookup.
 */
export async function getTreasuryBalances(
  transport: Transport,
): Promise<ChainBalance[] | null> {
  try {
    const { body } = await transport.request<BalancesByChainResponse>({
      method: "GET",
      path: "/api/balances/by-chain",
      expectStatuses: [200],
    });
    return body.balances ?? null;
  } catch {
    return null;
  }
}

File: python/canopy_ai/balances.py
"""
Per-chain treasury balance helper for the SDK's multi-chain routing path.

``canopy.fetch()`` calls this only when a 402 advertises ≥2 candidate rails
to pick the funded one. Single-candidate flows skip the round-trip and
just attempt the offer.
"""

from __future__ import annotations

from typing import TypedDict

from canopy_ai.transport import Transport


class ChainBalance(TypedDict):
    chainId: int
    chainName: str
    usdcBalance: str


def get_treasury_balances(transport: Transport) -> list[ChainBalance] | None:
    """
    Fetch the org treasury's per-chain USDC balance via
    ``/api/balances/by-chain``. Returns None on failure (treasury not
    provisioned, every chain RPC down, …) so callers fall back to the
    server's preference order — better to attempt the payment and let
    ``/api/sign`` surface a clear error than to wedge the request.
    """
    try:
        _, body = transport.request(
            "GET",
            "/api/balances/by-chain",
            expect_statuses=[200],
        )
    except Exception:
        return None
    if not isinstance(body, dict):
        return None
    balances = body.get("balances")
    if not isinstance(balances, list):
        return None
    return balances
